libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxVoiceApi.kt

```diff
@@ -0,0 +1,112 @@
+import java.io.File
+
+/**
+ * Implementation of [VoiceApi] allowing you to retrieve voice instructions.
+ * @property context Context
+ * @property accessToken String
+ * @property language String
+ */
+internal class MapboxVoiceApi(
+    private val context: Context,
+    private val accessToken: String,
+    private val language: String
+) : VoiceApi {
+
+    /**
+     * Given [VoiceInstructions] the method returns a [File] wrapped inside [VoiceCallback]
+     * @param voiceInstruction VoiceInstructions object representing [VoiceInstructions]
+     * @param callback VoiceCallback contains [VoiceState.VoiceFile]
+     */
+    override fun retrieveVoiceFile(voiceInstruction: VoiceInstructions, callback: VoiceCallback) {
+        val action = VoiceAction.PrepareVoiceRequest(voiceInstruction)
+        when (val result = VoiceProcessor.process(action)) {
+            is VoiceResult.VoiceRequest -> {
+                val mapboxSpeech = result.request
+                    .accessToken(accessToken)
+                    .language(language)
+                    .build()
+                mapboxSpeech.enqueueCall(object : Callback<ResponseBody> {
+                    override fun onResponse(
+                        call: Call<ResponseBody>,
+                        response: Response<ResponseBody>
+                    ) {
+                        val voiceAction = VoiceAction.ProcessVoiceResponse(response)
+                        when (val res = VoiceProcessor.process(voiceAction)) {
+                            is VoiceResult.Voice.Success -> {
+                                ThreadController.getMainScopeAndRootJob().scope.launch {
+                                    val voiceFile = generateVoiceFileFrom(res.data)
+                                    callback.onVoiceFileReady(VoiceState.VoiceFile(voiceFile))
+                                }
+                            }
+                            is VoiceResult.Voice.Failure -> {
+                                callback.onError(
+                                    VoiceState.VoiceError(res.error)
+                                )
+                            }
+                            is VoiceResult.Voice.Empty -> {
+                                callback.onError(
+                                    VoiceState.VoiceError("No data available")
+                                )
+                            }
+                        }
+                    }
+
+                    override fun onFailure(call: Call<ResponseBody>, t: Throwable) {
+                        callback.onError(
+                            VoiceState.VoiceError(t.localizedMessage ?: "Unknown")
+                        )
+                    }
+                })
+            }
+        }
+    }
+
+    /**
+     * Given the [Announcement] the method may cleanup any associated files previously generated.
+     * @param announcement
+     */
+    override fun clean(announcement: Announcement) {
```

Comment:
Should this be `VoiceInstructions` or is there a way for the caller to get the `Announcement`?

Reply:
https://github.com/mapbox/mapbox-navigation-android/blob/9d9590ccafcb89459af455a9fe70b84435ce81d4/test-app/src/main/java/com/mapbox/navigation/examples/core/VoiceActivity.kt#L137-L141

Reply:
should `clean` be `internal` then?

Reply:
Nah, it's up to the client to clean up the resources. Was thinking about doing this automatically but then the `SpeechPlayer` needs to know about the `SpeechApi` and wanted to avoid that coupling.

Reply:
> Nah, it's up to the client to clean up the resources. 

I'm not seeing why we want to make require the client to do this yet 🤔

Reply:
`SpeechPlayer` would only need to know about the `MapboxSpeechFileProvider`. Maybe that's a worthy tradeoff? I agree with @kmadsen that since we don't have caching mechanism, we should be cleaning up after ourselves automatically.

Reply:
But I also see that if someone would like to use the `SpeechPlayer` without our `SpeechApi`, then we would be removing their files which isn't ideal. The way the implementation is decoupled does make sense that users need to clean up manually, it just isn't convenient. We have an argument both ways I guess.

Reply:
Correct, I wanted to avoid that coupling.

Reply:
Also, we were discussing that it may be a use case for the clients that would like to keep the files ¯\\\_(ツ)\_/¯

Reply:
I don't think reusability is a valid scenario right now since we assign random IDs which reset with every instantiation of the SDK.